### PR TITLE
Fastd merge

### PR DESCRIPTION
**Pull request by rlei**
A proper pull request with all up to date patches.
- comments
- Doxygen cmake fixes for Android target
- other misc fixes

This pull request is based on the branch `fastd_upsteam` which is directly pulled from fastd git.

### CODE REVIEW

**Comment by rlei** [comment]
@NeoRaider could you please review this pull request with latest changes?

File: src/tuntap.c [hunk 1]
```diff
@@ -60,16 +59,15 @@ static const bool multiaf_tun = false;
 static const bool multiaf_tun = true;
 #endif
 
+#if defined(__ANDROID__) || defined(__linux__)
```
Comment (neocturne) [inline src/tuntap.c]: Is `__linux__` really not defined on Android?

File: src/tuntap.c [hunk 2]
```diff
@@ -60,16 +59,15 @@ static const bool multiaf_tun = false;
...
 
-#if defined(__linux__)
+/** Opens the TUN/TAP device helper shared by Android and Linux targets */
+void fastd_tuntap_open_linux(const char * dev_name) {
```
Comment (neocturne) [inline src/tuntap.c]: This should be `static` and lose its `fastd_` prefix.

**Comment by neocturne** [comment]
You mentioned you also wrote an Android build howto, it would be great to have that in `/doc` (as RST in the Sphinx docs would be best, but a text file in `/doc` would be fine for now too, as there isn't any other build documentation in the Sphinx so far...)

Also, I'd prefer to merge this squashed and rebased, as the history contains many small fixups.

Only two comments left :) I think after these are fixed, we're ready to go and I can merge the Android support. If you want, you can also add your name and mail address to the source files you edited by more than a few lines, and the toplevel `COPYRIGHT` file. I guess I should also add an `AUTHORS` file then...

File: src/tuntap.c (hunk 1 above)
Comment (rlei) [inline src/tuntap.c]: yep, my mistake.

**Comment by rlei** [comment]
@NeoRaider I wrote a simple [README.md](https://github.com/rlei/fastd-android/blob/master/README.md) and a rough [build script](https://github.com/rlei/fastd-android/blob/master/build-fastd-android.sh) which also pulls/builds libuecc & libsodium, because getting everything in right place can be tricky for new comers to Android and/or fastd. They were not included in current `fastd_merge` branch.

Can you have a first look and give your comments? I can move both to `doc` and rename `README.md` to something like `REAMDE.android.md`. They'll be included in the squashed commit after being reviewed.

**Comment by neocturne** [comment]
I think moving the README and the build script to `doc/android` would be okay. I've only looked through the build script roughly, but it should be fine to merge like this if you fix the references in the comments to point at the official fastd repository.

**Comment by rlei** [comment]
@NeoRaider all changes squashed and pushed to `fastd_merge` branch now. Thanks for your reviewing :)

**Comment by neocturne** [comment]
I found one more issue when looking through the changes once more:

When running fastd on rooted android without `--android-tun`, you use the normal TUN/TAP creation code; I think it would make sense to allow user, group and PID file settings as well on rooted android.

I'd suggest renaming the parameter to `--android-integration`, which should enable the following features:
- Receive TUN fd over control socket
- Protect sockets
- Reset all peers on SIGHUP
- Disable user, group and pidfile support

In conclusion, without `--android-integration`, fastd would behave exactly like it does on any other Linux system. Does this make sense?

One last thing is the commit timestamp, which is currently January 5th. Please recreate the commit with `--amend --reset-author` to update the timestamp, so the git history is a bit clearer.

**Comment by rlei** [comment]
In fact the pidfile support was disabled partially as a quick workaround because Android Bionic (libc) doesn't provide `dprintf`. Would you like me to change it to `snprintf` + `write`? or maybe `fopen`/`fwrite`?

**Comment by neocturne** [comment]
`fopen`+`fprintf` seems like the most natural solution to me.

**Comment by rlei** [comment]
yep I meant to say `fprintf`

**Comment by neocturne** [comment]
Ah, and you could also change `cmake/config.cmake` to not even show the `ENABLE_SYSTEMD` option on Android as that uses `dprintf` as well and doesn't make any sense on Android.

**Comment by rlei** [comment]
ok will do

**Comment by rlei** [comment]
amended and pushed.

File: src/fastd.c [hunk 3]
```diff
@@ -256,11 +266,12 @@ static inline void write_pid(void) {
 		pr_error_errno("can't write PID file: open");
 		goto end;
 	}
+	FILE *f = fdopen(fd, "w");
```
Comment (neocturne) [inline src/fastd.c]: You can just use normal `fopen` instead of `open` and `fdopen`

**Comment by neocturne** [comment]
What about my other suggestions regarding the `--android-integration` option?

**Comment by rlei** [comment]
`--android-integration` is implemented in last push too.

File: src/fastd.c (hunk 3 above)
Comment (rlei) [inline src/fastd.c]: I just want to make sure the `0666` mode is preserved as is. Is it ok just to drop it and use `fopen` instead?

**Comment by neocturne** [comment]
I also suggested to do the following only when `--android-integration` is set:
- Protect sockets
- Reset all peers on SIGHUP
- Disable user and group support

At the moment, these things are done unconditionally when `__ANDROID__` is set. Do you think my suggestions make sense? I'd like fastd to behave as much like on other Linux systems as possible when `--android-integration` is not set.

File: src/fastd.c (hunk 3 above)
Comment (neocturne) [inline src/fastd.c]: The `fopen` manpage says the following:

   Any created files will have mode S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH (0666), as modified by the process's umask value (see umask(2)).

**Comment by rlei** [comment]
Oh sorry I overlooked your suggestions:
- Protect sockets is already implemented in this way. The `protect_socket` call directly returns true and does nothing when the integration flag is not set.
- Due to the nature of Android that network connectivity may change all the time (cellular -> WiFi -> cellular etc), I feel reseting peers on certain signal is still needed even for rooted device, otherwise fastd will never get the chance to recreate the socket unless it's restarted. On the other hand, this will only affect very few users...so it'll be your call :)
- I thought user/group support is along with pidfile but it's not. Will update it too.

It's a bit late for my timezone now so will update the pull request tomorrow...

**Comment by neocturne** [comment]
Hmm, what exactly sends the SIGHUP on connectivity changes? I find it a bit problematic that SIGHUP is used both for peer list reloads and connection resets, so that leaves you no way to reload the peer configs without resetting even unchanged peer connections.

File: src/fastd.c (hunk 3 above)
Comment (rlei) [inline src/fastd.c]: ok will change to `fopen`.

**Comment by rlei** [comment]
My currently work-in-progress Android frontend to fastd will send `SIGHUP`. Any other frontends that follow the very simple control socket protocol can do so. I'll also push the frontend to github once basic features are implemented.

(non rooted) Android allows only one TUN being opened so it should not be a problem to use `SIGHUP` to signal resetting the peer (because when changing/reloading peer config, it is the only one config).

The tricky part is still the use case of running fastd on rooted Android. I'm not sure if anyone would actually run multiple VPN on Android phones, even when rooted.

That said, I was actually using `SIGUSR1` (which was for reporting fastd status but recently got dropped in favor of status socket IIRC) to signal the resetting in the first place, but changed to `SIGHUP` in hope that reloading config(s) would just do the resetting work. It turned out not so I added the forced resetting.

But using a different signal is also making the Android port different to other ports. Looks like the simplest way is just as you said, just don't reset peers on `SIGHUP` without `--android-integration`. If someone really wants to run fastd on a phone without the flag, he'll have to restart fastd on connectivity changes.

**Comment by neocturne** [comment]
As a way to reset all connections is generally useful, I've now added such a feature (and use `SIGUSR2` for that), so `--android-integration` should be changed to not affect the `SIGHUP` handling.

**Comment by rlei** [comment]
Great, will merge later today.

**Comment by rlei** [comment]
OK now yet another issue about the user/group setting feature. I now recall that this part was disabled due to:
- Android Bionic has no impl of `getpwnam_r`
- since Android has no /etc/passwd at all, the impl of `getpwnam` is [searching in an internal table](http://androidxref.com/5.0.0_r2/xref/bionic/libc/bionic/stubs.cpp#210) that consists preallocated UIDs for installed applications plus system services. Each end user on a phone has internally a "UID space" of 100,000 UIDs, and all apps installed to his account will be assigned a unique Linux UID within this space. Same installed apps but under different user accounts on a same phone will receive different UIDs in different user UID spaces.

So it hardly serves any meaningful purpose to specify user/group, I would suggest just keep current modification that simply `exit_error` for user/group setting, regardless of whether `--android-integration` is set or not.

**Comment by rlei** [comment]
Changes in last push:
- merged and rebased to fastd master
- `open` + `fdopen` => `fopen`
- revert force reseting peers change
- only disable PKTINFO in send.c when `--android-integration` specified

File: src/fastd.c [hunk 4]
```diff
@@ -256,16 +266,16 @@ static inline void write_pid(void) {
...
 		goto end;
 	}
 
-	if (dprintf(fd, "%u", (unsigned)getpid()) < 0)
+	if (fprintf(f, "%u", (unsigned)getpid()) < 0)
 		pr_error_errno("can't write PID file: dprintf");
 
-	if (close(fd) < 0)
+	if (fclose(f) < 0)
 		pr_warn_errno("close");
```
Comment (neocturne) [inline src/fastd.c]: The three error messages must be adjusted to the changed function calls.

File: src/send.c [hunk 5]
```diff
@@ -117,7 +121,14 @@ static void send_type(const fastd_socket_t *sock, const fastd_peer_address_t *lo
 	msg.msg_control = cbuf;
 	msg.msg_controllen = 0;
 
+#ifdef __ANDROID__
+	/* PKTINFO will mess with Android VpnService.protect(socket) */
+	if (!conf.android_integration) {
+		add_pktinfo(&msg, local_addr);
+	}
+#else
```
Comment (neocturne) [inline src/send.c]: I think this check should be moved to `add_pktinfo` to avoid the `#else` branch.

File: src/fastd.c [hunk 6]
```diff
@@ -256,16 +266,16 @@ static inline void write_pid(void) {
 			pr_debug_errno("seteuid");
 	}
 
-	int fd = open(conf.pid_file, O_WRONLY|O_CREAT|O_TRUNC, 0666);
-	if (fd < 0) {
+	FILE * f = fopen(conf.pid_file, "w");
```
Comment (neocturne) [inline src/fastd.c]: Just a nitpick: there should be no space after the `*`.

**Comment by neocturne** [comment]
Okay, the user/group handling is fine like this then (I might refactor it later, but that's no reason not to merge it like this)

**Comment by rlei** [comment]
Fixed and pushed again.

**Comment by neocturne** [comment]
Very nice! If you're okay with it, I'll merge it like this.

**Comment by rlei** [comment]
Please go ahead with the merge - thank you for the review and patience :)

**Comment by neocturne** [comment]
Merged - thanks for your contribution :)
